### ISSUE TITLE
No trusted server host keys available

### ISSUE THREAD

**Issue by cloudtrends** [ronf/asyncssh#52]
hi all,

I just want run hello world example of asyncssh . 

My python version is : 3.5 

> python3.5  simple_client.py 
SSH connection failed: Disconnect Error: No trusted server host keys available

I had already run  simple_server.py  from another terminal .

After check source code of simple_client.py , it just try to connect localhost .

so , where is the location for  trusted server host keys for localhost ?

thanks in advance .

**Comment by redixin** [comment]
Hi. Usually trusted hosts are listed in ~/.ssh/known_hosts

This file may be absent if you never used ssh client at all. Just guessing: asyncssh is trying to open nonexistent file and failing with error.

**Comment by cloudtrends** [comment]
@redixin   thanks 

I checked that ,  the file  ~/.ssh/know_hosts exist  , and file permission is :  

>[xxx@xxx examples]$ ls -lth  ~/.ssh/known_hosts 
-rw-r--r--. 1 xxx xxx 2.2K Mar  9 12:10 /home/xxx/.ssh/known_hosts

after digging into the source code , I found that  ,   the error message " No trusted server host keys available "  only occurred in file ./asyncssh/connection.py .

but after I modified the error message  (for debug purpose ),   

when I run 
> python3.5 simple_client.py

the console still show the error message :   SSH connection failed: Disconnect Error: No trusted server host keys available . ( this mean the error message is not from asyncssh itself .  )


My question is  how can I add  localhost into  ~/.ssh/known_hosts file ?

**Comment by redixin** [comment]
You can add localhost to know_hosts by just connecting to it by ssh:

    ssh localhsot

and accept ssh host key

**Comment by ronf** [comment]
That's right - the default behavior is to look in ~/.ssh/known_hosts, so using the regular ssh client to connect to a host is an easy way to get a key added there. However, you can also specify an alternate value for known_hosts in the AsyncSSH call to open the connection if you like. More information on what values you can pass in can be found at:

http://asyncssh.readthedocs.org/en/latest/api.html#known-hosts

If you don't want to create an entry in known_hosts format and you have access to the server's public key, you can just pass that in directly by using the variation where you pass in a tuple containing the host keys to trust, CA keys to trust, and revoked keys to no longer trust. So, for instance, if the server's host public key was available in the file 'server_key.pub', you could call:

```python
    yield from asyncssh.connect('localhost', known_hosts=(['server_key.pub'], [], []))
```

This says to trust server_key.pub as a host key and specifies empty lists for trusted CA keys and revoked keys.

**Comment by ronf** [comment]
Regarding the error you're getting, it is being returned by asyncssh/connection.py. If you modified that and still got the original version of the error, you probably had another instance of AsyncSSH installed that was being used instead of the one you modified. The error means that there was no match for the host key presented by the server in whatever known_hosts file you were using.

**Comment by cloudtrends** [comment]
@redixin   follow your way , I successed add localhost into my   ~/.ssh/known_hosts file 
@ronf    you are right , I had another version of asyncssh which installed by pip3 , and I uninstall it .

after changing the simple_client.py code 
from 
> asyncssh.connect('localhost'))   
to  
> asyncssh.connect('localhost', known_hosts=(['/home/xxx/.ssh/id_rsa.pub'],[],[] ))) 

the error message showed as below:
> SSH connection failed: Disconnect Error: Untrusted server host key

btw, this time , I run the simple_client.py in pycharm IDE instead of run directly from console .

**Comment by ronf** [comment]
The "id_rsa.pub" key you listed here is a client SSH key, not a server SSH key. You'd need to specify the server SSH key, which would typically be located in something like "/etc/ssh/ssh_host_rsa_key.pub" on the target machine. In the case of connecting to localhost, you might be able to point at this file directly, but in the more general case where you are ssh'ing to a remote machine, you'd need to manually copy that key to your local machine somewhere that your client code can read it from.

**Comment by denismakogon** [comment]
Hello.

I'm facing the same problem as @cloudtrends.

So, i do slightly different thing. I'm using OpenStack to create VM with SSH key in it (it was delivered to /home/ubuntu/.ssh/authorized_keys). And once instance is up i want to SSH into it and run post-install script. But asyncssh keeps telling me - Untrusted server host key.

IIRC in ssh client config you can specify to disable restrict host key check, if there's something similar in asyncssh?

Here's my code https://github.com/aiorchestra/aiorchestra-asyncssh-plugin/blob/master/asyncssh_plugin/tasks.py#L41-L64, would you mind to tell me what am i doing wrong?

**Comment by ronf** [comment]
The "known_hosts=None" that you have in there should disable the server host key check on the client. There's an extra trailing comma there that I wouldn't have thought was legal, but it doesn't look like that's causing the problem.

Can you try simplifying your example and see if you can still reproduce it?

The error you're getting is a curious one. It's saying that there was a known_hosts entry for the host you are trying to connect to, but it's not the right host key (as opposed to it finding no host key). Are you sure you are actually running the version of the code linked above? It looks more like you are passing in a known_hosts file, but it contains the wrong key.

**Comment by denismakogon** [comment]
@ronf thanks for response, actually, i've used "known_hosts=None" and it worked for me, and i think it is worth to write doc for how to use connection without strict host key checks, because it is not that obvious for asyncssh newbies. But it is interesting, because according to those docs you've made it would work if i pass imported public key to known hosts (similar as you've told to @cloudtrends) but it doesn't work for me either.
I also tried to import known hosts from string:

      asyncssh.import_known_hosts("{0} {1}".format(host, public_key))

and then pass it to know_hosts parameter.

So, can we improve connection implementation to have some sort of ssh client config that would explicitly define all required options to establish connection from host to host (similar to .ssh/ssh_config at client site) ?

**Comment by ronf** [comment]
What format is "public_key" in this case? It needs to be in OpenSSH format (single line, with a key type like "ssh-rsa" or "ecdsa-sha2-nistp256" followed by a space followed by base64-encoded key data. Alternately, you can pass in the keys you want to trust using the list format I described above without calling import_known_hosts(). In that case, any of the supported public key formats can be used to hold the key. If you want to pass in raw key data without writing it to a file, just make sure the argument you pass in the lists is a "bytes" object rather than a "str" -- bytes values are passed to import_public_key() and str values are treated as filenames and passed to read_public_key().

Regarding the config file, I've resisted adding support for OpenSSH's .ssh/config file so far, because there are many minor differences in the options supported by OpenSSH vs. AsyncSSH, even though they interoperate quite well with one another. I could invent my own config file syntax instead, but just about everything you'd be able to do with a config file could easily be expressed in Python itself and that's a lot more flexible than any config file format I might invent.

Basically, everything that would appear in a config file can be passed in as arguments to asyncssh.create_connection() or asyncssh.connect(). If you want to repeatedly pass in a particular set of arguments, you can build your own wrapper function that does this so you don't need to repeat the arguments that don't change. That wrapper could also automatically adjust the arguments on something like a per-host basis (or based on any other arguments you pass in) by including appropriate conditionals.
